### PR TITLE
flux-shutdown: add --gc garbage collection option

### PR DESCRIPTION
**Pull request by garlick**
This adds some logic to rc1 and rc3 to enable offline garbage collection of a system instance using `flux-dump(1)` and `flux-restore(1)`.  If requested by `flux shutdown --gc`, a dump is produced in ${statedir} during shutdown with a RESTORE symlink pointed to it.  On startup, if RESTORE exists, the current backing store is truncated and content is restored from the archive file.  Then the RESTORE symlink is removed.

For example on my test system:
```
$ sudo flux shutdown --gc
flux-shutdown: shutdown will dump KVS (this may take some time)
broker.info[0]: cleanup.0: flux queue stop --quiet Exited (rc=0) 0.0s
broker.info[0]: cleanup.1: flux job cancelall --user=all --quiet -f --states RUN Exited (rc=0) 0.0s
broker.info[0]: cleanup.2: flux queue idle --quiet Exited (rc=0) 0.0s
broker.info[0]: cleanup-success: cleanup->shutdown 0.104717s
broker.info[0]: children-none: shutdown->finalize 0.179421ms
broker.info[0]: rc3.0: dumping content to /var/lib/flux/dump-20220425_174130.tgz
broker.info[0]: rc3.0: /usr/local/etc/flux/rc3 Exited (rc=0) 0.6s
broker.info[0]: rc3-success: finalize->goodbye 0.578407s
$ sudo ls -l /var/lib/flux
total 39684
-rw-r--r-- 1 flux flux  1073152 Apr 25 17:41 content.sqlite
-rw-r--r-- 1 flux flux    48183 Apr 25 17:41 dump-20220425_174130.tgz
-rw-r--r-- 1 flux flux 39510016 Apr 24 18:09 job-archive.sqlite
lrwxrwxrwx 1 flux flux       24 Apr 25 17:41 RESTORE -> dump-20220425_174130.tgz
$ sudo systemctl start flux
$ sudo ls -l /var/lib/flux
total 40896
-rw-r--r-- 1 flux flux     4096 Apr 25 17:42 content.sqlite
-rw-r--r-- 1 flux flux  2307232 Apr 25 17:42 content.sqlite-wal
-rw-r--r-- 1 flux flux    48183 Apr 25 17:41 dump-20220425_174130.tgz
-rw-r--r-- 1 flux flux 39510016 Apr 24 18:09 job-archive.sqlite
```

It's also possible to checkpoint/restart an instance with:
```
$ flux shutdown --dump=foo.tar.bz2
$ flux start -o,-Scontent.restore=foo.tar.bz2
```

although this is only practical if R hasn't changed at the moment.

Marking a WIP as I wanted to get feedback on the approach before writing tests.

I had thought maybe garbage collection could be automated somehow by tracking some metric that could be used as an indicator of the need.  However, maybe getting a bit of experience with doing it manually first makes sense.

### CODE REVIEW

**Comment by garlick** [comment]
I went ahead and added test coverage here and have been testing this on my home test system, so removing the WIP.  I'm still open to doing this another way if people have better ideas!

**Comment by grondo** [comment]
Sorry, I haven't had time to take a peek at this. I can't think of a any better interface, given that garbage collection must occur as part of a dump/restore. I can't remember, does GC happen just as a natural result of the restore? 

As a more general point (and I guess unrelated to this PR), I'm a little worried there will be confusion when to use `flux shutdown` vs systemd for stopping a Flux system instance. I'm not sure there are many systemd services that are stopped via a different command.

**Comment by garlick** [comment]
Yes a dump/restore walks the KVS metadata starting with the last root hash, so when it is restored, all the unreferenced data is left behind.  In addition, the archive only contains "files", unlike a file system archive created with tar (for example).  So empty job directories are removed also.

One can still run `systemctl stop flux` on the rank 0 onde to shut down the instance.  And one could trigger a dump/restore as part of that by manually setting the `content.dump` attribute to `auto`.  However I didn't want to encourage that because then the dump would be subject to the systemd `TimeoutStopSec`  and would risk getting killed before the dump is complete.  Hence adding the option to `shutdown` so it's tied to that way of bringing down the instance.  I anticipate that we will add other things to `flux-shutdown(1)` like options to stop the queue and let running jobs complete, or to shut down in the future.  So maybe it will become a natural way to stop flux.

One other weakness I see here is the dump files aren't removed and will pile up after a while.  I was vaguely thinking that this could be valuable if we wanted to revert to a previous checkpoint if the db was corrupted or whatever, and that sys admins could manage the dump files with log rotation tools.  Is that reasonable?

I guess the other question - is `--gc`  an annoyingly terse option?  It's just a shorthand for `--dump=auto`.  Should we just go with that?  The purpose of `--gc` was to provide an option that matched the desired end effect (garbage collecting the KVS).  I'm open to better  names.

**Comment by grondo** [comment]
Ah, thanks for that refresher, that was helpful. Given the above, this approach seems just fine IMO. I like the idea of extending shutdown semantics in the future. Also, if dumpfiles tend to accumulate, maybe logrotate or systemd-tmpfiles could be configured to automatically clean things up?

**Comment by grondo** [comment]
And FWIW, I don't have a problem with `--gc`. It is close enough to `git gc` where I understand what it is meant to do.

**Comment by garlick** [comment]
Pushed a tmpfiles.d config file and moved "auto" dumps to `$statedir/dump` since that made the tmpfiles rule easier to write.

Tested on my test system instance by running `flux shutdown --gc` a few times and reducing the age setting for dump files, then observing that they were purged with `systemd-tmpfiles --clean`

**Comment by garlick** [comment]
Repushed with reference to #258

File: src/modules/content-sqlite/content-sqlite.c [hunk 1]
```diff
@@ -807,7 +807,7 @@ static int process_args (struct content_sqlite *ctx,
             *truncate = true;
         }
         else {
-            flux_log_error (ctx->h, "Unknown module option: '%s'", argv[i]);
+            flux_log (ctx->h, LOG_ERR, "Unknown module option: '%s'", argv[i]);
```
Comment (chu11) [inline src/modules/content-sqlite/content-sqlite.c:810]: perhaps should stylize do a similar change in `content-files`?  (`content-files` sets errno = EINVAL to make it not as bad)

Comment (chu11) [reply src/modules/content-sqlite/content-sqlite.c:810]: oh and I guess w/ `content-s3` too (given follow up commit to this one)

File: t/t0018-content-files.t [hunk 2]
```diff
@@ -73,6 +73,10 @@ kvs_checkpoint_get() {
 # Tests of the module by itself (no content cache)
 ##
 
+test_expect_success 'content-files module load fails with unknown option' '
+	test_must_fail flux module load content-files notoption
+'
+
```
Comment (chu11) [inline t/t0018-content-files.t:79]: nit, should this test be a different commit? not really related to ` content-files: add truncate module option `

File: etc/rc3 [hunk 3]
```diff
@@ -45,7 +45,25 @@ modrm all kvs
...
+            dumplink="${statedir:-.}/dump/RESTORE"
+        fi
+        echo "dumping content to ${dumpfile}"
+        if flux dump --quiet --checkpoint ${dumpfile}; then
+            test -n "$dumplink" && ln -s $(basename ${dumpfile}) ${dumplink}
+        else
+            exit_rc=1
+        fi
+    fi
+    flux module remove ${backingmod} || exit_rc=1
```
Comment (chu11) [inline etc/rc3:66]: should use `modrm`?

File: etc/rc1 [hunk 4]
```diff
@@ -16,7 +12,34 @@ modload() {
...
+            if test -h "${dumplink}"; then
+                dumpfile=$(readlink -f ${dumplink}) || :
+            else
+                dumpfile=""
+                dumplink=""
+            fi
+        fi
+    fi
+    if test -n "${dumpfile}"; then
+        flux module load ${backingmod} truncate
```
Comment (chu11) [inline etc/rc1:33]: use modload for consistency?

File: etc/rc1 [hunk 5]
```diff
@@ -16,7 +12,34 @@ modload() {
...
+        echo "restoring content from ${dumpfile}"
+        flux restore --quiet --checkpoint ${dumpfile}
+        if test -n "${dumplink}"; then
+            rm -f ${dumplink}
+        fi
+    else
+        flux module load ${backingmod}
```
Comment (chu11) [inline etc/rc1:40]: use modload for consistency?

File: etc/rc1 [hunk 6]
```diff
@@ -16,7 +12,34 @@ modload() {
...
+            fi
+        fi
+    fi
+    if test -n "${dumpfile}"; then
+        flux module load ${backingmod} truncate
+        echo "restoring content from ${dumpfile}"
+        flux restore --quiet --checkpoint ${dumpfile}
+        if test -n "${dumplink}"; then
+            rm -f ${dumplink}
+        fi
```
Comment (chu11) [inline etc/rc1:38]: should only remove the link if the restore is successful?

**Review by chu11 — COMMENTED** [review]
overall lgtm, just a few comments / nits I found

File: etc/rc3 (hunk 3 above)
Comment (garlick) [reply etc/rc3:66]: `modrm` tests $RANK before running `flux module remove`, but this is within a block that is already conditional on $RANK.   I thought it kind of weird to trade a straightforward one-liner for a function call to do same when it was necessary to repeat the rank constraint.  Does that make sense?

File: etc/rc1 (hunk 4 above)
Comment (garlick) [reply etc/rc1:33]: same deal here

File: etc/rc1 (hunk 5 above)
Comment (garlick) [reply etc/rc1:40]: ibid

File: etc/rc1 (hunk 6 above)
Comment (garlick) [reply etc/rc1:38]: Yes but the rc1 shebang is `#!/bin/bash -e` so the script aborts before the remove if the restore is unsuccessful.

**Comment by garlick** [comment]
Just pushed fixes based on @chu11's comments, and also rebased on current master.

File: etc/rc3 (hunk 3 above)
Comment (chu11) [reply etc/rc3:66]: ahh that makes sense, you have the `rank == 0` check above this.

**Review by chu11 — APPROVED** [review]
LGTM!

**Comment by garlick** [comment]
Thanks!  I'll set MWP.
